[ADD] ai_oca_bridge

etobella opened the pull request:
This module is used to create a bridge between Odoo and other AI systems
like n8n.

Define a bridge and it will be shown automatically in the chatter of the model.

TODO:

- [x] Finish Documentation
- [x] Add tests for JS
- [x] Add tests for Python
- [ ] Forward port to 17-18

https://github.com/user-attachments/assets/906b95fe-3810-4e20-adf7-d323ac3b494f

This example only shows the communication. We could do a lot of things from this point.

The only limit is our imagination :wink: 

@ValentinVinagre

etobella committed:
[ADD] ai_oca_bridge

HaraldPanten reviewed (changes requested):
You should change the name of the first commit.

ai_bridge instead of ia_bridge

ValentinVinagre commented on ai_bridge/readme/CONTEXT.md:
```diff
@@ -0,0 +1,12 @@
+Right now, there are 2 different approaches for AI integration with Odoo:
+
+1- Make everything inside Odoo.
```
change "1.-" and "2.-" for numered list "#. "

![image](https://github.com/user-attachments/assets/e099929c-32b2-477c-9de3-20f29cf682ed)

etobella commented:
The module is working, and tests are there. I think this can opt as v1 of this module and then, we can start working from this module.

I am open to add new ideas to ROADMAP, but I would prefer to keep starting logic simple.

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,130 @@
+    active = fields.Boolean(default=True)
+    description = fields.Text()
+    model_id = fields.Many2one(
+        "ir.model",
```
Non-persistent models should be filtered out, as well as other security-related models like "res.users, res.groups,..."
How do you see it?

transient=False

etobella replied:
transient for sure. about ignoring other models, I am not so sure

ValentinVinagre replied:
Malicious configurations could be possible. I'm not entirely sure if it would be better to have a check mark in the models to indicate "safe" ones and then have the user expand said configuration. What do you think?

etobella replied:
well, this menu is only accessible and editable by system adminsitrators :thinking:

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,130 @@
+        compute="_compute_execution_count",
+    )
+    url = fields.Char(
+        string="URL",
```
Do we validate the URL?
tools.validate_url()

etobella replied:
No, we don't validate it. However, tools.validate_url will only add http at the begining. I am not sure if this fulfills ours needs :thinking:

ValentinVinagre commented on ai_bridge/static/src/components/chatter_topbar/chatter_topbar.xml:
```diff
@@ -0,0 +1,54 @@
+        owl="1"
+    >
+        <xpath
+            expr="//div[hasclass('o_ChatterTopbar_rightSection') and not(@id='mobileChatterTopbar')]/button[last()]"
```
On a clean install of CE 16.0 this file gives an error blocking the system.
Is it possible that the responsive web dependency is missing?
```
odoo.addons.base.models.ir_qweb.QWebException: Error while render the template
odoo-1        | ValueError: Element '<xpath expr="//div[hasclass(&#39;o_ChatterTopbar_rightSection&#39;) and @id=&#39;mobileChatterTopbar&#39;]/button[last()]">' cannot be located in parent view
odoo-1        | Template: web.webclient_bootstrap
odoo-1        | Path: /t/t/t[1]/t[4]
odoo-1        | Node: <t t-call-assets="web.assets_backend" t-css="false"/>
odoo-1        | Compiled code:
odoo-1        | code = None
odoo-1        | template = 'web.webclient_bootstrap'
```

etobella replied:
there is a bug with web_responsive and I made something to make it work, but it was even worse. I will check it

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,130 @@
+        return True
+
+    def _prepare_payload(self, record, **kwargs):
+        vals = {}
```
please add: self.ensure_one()

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,130 @@
+        for record in self:
+            record.field_ids = False
+
+    @api.depends("field_ids", "model_id")
```
Do you think it's necessary to reset the sample value every time a field is specified? You might need to add an extra field, and that would eliminate the configuration already set up. I would only delete it if the model changes. What do you think?

etobella replied:
this payload is a sample that will be used for testing. I don't see the problem.

Also, this paayload will never be used to connect. It is useful only from visibility perspective.

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,130 @@
+        self.ensure_one()
+        record = self.env[res_model].browse(res_id).exists()
+        if record:
+            self.env["ai.bridge.execution"].create(
```
Do you think it's correct to indicate whether the connection was successful? This way, quick feedback can be provided without having to navigate to the execution log. Another way would be to return the execution log to a form view to review the execution directly. What do you think?

etobella replied:
Well, it depends on the external system. For example, n8n doesn't send you the id of the execution. Making it hard to evaluate it.

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,130 @@
+    _name = "ai.bridge"
+    _description = "Ai Bridge Configuration"
+    _order = "sequence, id"
+
```
It would be necessary to make it multi-company. Do you think it is correct for this version?

etobella replied:
Sure

ValentinVinagre commented on ai_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,70 @@
+        ondelete="cascade",
+    )
+    res_id = fields.Integer(required=True)
+    model_id = fields.Many2one(
```
Do you find it necessary to delete the executions of a deleted model?

etobella replied:
Otherwise it cannot be required....

ValentinVinagre commented on ai_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,70 @@
+        required=True,
+        ondelete="cascade",
+    )
+    payload = fields.Json()
```
readonly=True ?

ValentinVinagre commented on ai_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,70 @@
+        ondelete="cascade",
+    )
+    payload = fields.Json()
+    result = fields.Text()
```
readonly=True?

ValentinVinagre commented on ai_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,70 @@
+        payload = self.ai_bridge_id._prepare_payload(
+            self.env[self.model_id.model].browse(self.res_id), **kwargs
+        )
+        response = requests.post(
```
it's not necessary the "try catch" because it is incorporated in the requests library correct?

etobella replied:
yes, however, I made a raise_for_status later :smile:

ValentinVinagre commented on ai_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,70 @@
+        self.payload = payload
+
+    def _execute_kwargs(self, timeout=False, **kwargs):
+        result = {}
```
self.ensure_one() ?

etobella replied:
It is not really necessary for private functions.... anyway, I can do that.

ValentinVinagre commented on ai_bridge/security/ir.model.access.csv:
```diff
@@ -0,0 +1,4 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_ai_bridge_user,ai_bridge.user,model_ai_bridge,base.group_user,1,0,0,0
```
A registration rule would be missing depending on the configured company

etobella replied:
I understand you ask for company permissions. that makes sense and I will add it.

ValentinVinagre commented on ai_bridge/__manifest__.py:
```diff
@@ -0,0 +1,25 @@
+    "license": "AGPL-3",
+    "author": "Dixmit,Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/ai",
+    "depends": ["mail"],
```
please add 
```
development_status": "Beta
```
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/oca_module_lifecycle_development_status.rst#2alpha-and-beta-modules

etobella replied:
Alpha or Beta?

ValentinVinagre replied:
I think **Beta** -> Pre-production quality but with potential instability .

Documentation:

- The **Alpha** stage is not required, but it may be useful to signal that the module is not expected to work as planned yet (or at all). It is in the early stages of design and construction, and trying it on Runboat is pointless

- Changing the development status to **Beta** is a way to signal reviewers and functional testers that the module should be usable enough to be tested.

ValentinVinagre reviewed (commented):
small review 😄 , feel free to expand the roadmap.
It is currently not possible to install the module in 16CE.

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,143 @@
+        "res.groups",
+        help="User groups allowed to use this AI bridge.",
+    )
+    field_ids = fields.Many2many(
```
Would the domain be missing depending on the chosen model? It currently displays all the system fields. Would it be possible to navigate it like a system data export?

etobella replied:
Doing what you ask for makes it much more complex. I would avoid that as a first approach.

ValentinVinagre replied:
I completely agree.
I see a bit of the problem in the following configuration:
I want to transfer sales line data from a "sale.order." If a domain is set to only display the fields of the chosen template, those fields won't be selectable. A second connection would have to be made from the external system to retrieve that information.
I'm not sure if I'm explaining this correctly.
If you agree, add this limitation to the ROADMAP.

ValentinVinagre commented on ai_bridge/views/ai_bridge.xml:
```diff
@@ -0,0 +1,103 @@
+                        />
+                        <field
+                            name="auth_username"
+                            attrs="{'invisible': [('auth_type', '!=', 'basic')]}"
```
mandatory when select auth_type = basic?

ValentinVinagre commented on ai_bridge/views/ai_bridge.xml:
```diff
@@ -0,0 +1,103 @@
+                        />
+                        <field
+                            name="auth_password"
+                            attrs="{'invisible': [('auth_type', '!=', 'basic')]}"
```
mandatory when select auth_type = basic?
password="True" ?

ValentinVinagre commented on ai_bridge/views/ai_bridge.xml:
```diff
@@ -0,0 +1,103 @@
+                        <field name="auth_type" />
+                        <field
+                            name="auth_token"
+                            attrs="{'invisible': [('auth_type', '!=', 'token')]}"
```
mandatory when select auth_type = token?
password="True"?

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,143 @@
+    _name = "ai.bridge"
+    _description = "Ai Bridge Configuration"
+    _order = "sequence, id"
+
```
I think it would be a good idea to add the following statuses:

- draft: The user can configure all fields without problems, but it cannot be used in the widget, only test execution from the same ai_bridge.
- active/validated: The user cannot change important configuration parameters and this is taken into account when displaying the action in the user widget.
- cancel: The user has discarded this action, the configuration should not be able to be changed or executed.

What do you think?

etobella replied:
Mmmh, I don't like it. I would prefer to make a field called usage. By default we will have usage none and usage widget, This way it is cleaner and we can hide them or test them easily. WDYT?

ValentinVinagre replied:
I think it's fine; it's about having something to control its appearance while it's being configured. Another idea would be to assign it an appropriate permission group when creating it, and that's it, but I think it's better to have an asset field or something similar.

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,143 @@
+    )
+    name = fields.Char(required=True)
+    active = fields.Boolean(default=True)
+    description = fields.Text()
```
This information could be displayed as help for the execution widget action to indicate to the user that they are performing that action. What do you think?
or another "user info" field. The fields visible to users should be made translatable.

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,143 @@
+    execution_count = fields.Integer(
+        compute="_compute_execution_count",
+    )
+    url = fields.Char(
```
mandatory field?

etobella replied:
I didn't made it mandatory because I don't know if it makes sense to use it in other systems

ValentinVinagre replied:
How do you see adding a "connection" option? Currently, it could be indicated by URL, and if necessary, it could be expanded.
Although it could be done later.

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,143 @@
+        string="Authentication Type",
+        help="The type of authentication used to connect to the external AI system.",
+    )
+    group_ids = fields.Many2many(
```
Do you think it's appropriate to leave it blank? The default could always be "internal user," a required field, or the "group_system" group.
How do you see it?

etobella replied:
I don't see the problem, is like company_id False, that is open for all. The same happens with rules.

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,143 @@
+    _name = "ai.bridge"
+    _description = "Ai Bridge Configuration"
+    _order = "sequence, id"
+
```
I would add a chatter to save notes and changes, do you see it correct?

ValentinVinagre commented on ai_bridge/static/src/components/chatter_topbar/chatter_topbar.xml:
```diff
@@ -0,0 +1,41 @@
+            hotkey="'i'"
+            showCaret="true"
+        >
+            <t t-set-slot="toggler">
```
I think the action in the right corner will cause misclicks to occur when tracking the record. Would you consider moving it to the far left? The dropdown may need to be adjusted if there are long names; being on the right (current) side eliminates this problem.

![image](https://github.com/user-attachments/assets/3badd6ff-4dea-4b30-87a7-198cedb02026)

ValentinVinagre commented on ai_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,143 @@
+        for record in self:
+            record.field_ids = False
+
+    @api.depends("field_ids", "model_id")
```
I think the "description" field is missing in the depends

etobella replied:
Why? It is never used?

ValentinVinagre replied:
As I saw in the sample, I thought it was used.
![image](https://github.com/user-attachments/assets/2ffcba62-450d-4be9-972c-878b41402275)

ValentinVinagre commented on ai_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,82 @@
+            **self._execute_kwargs(**kwargs),
+        )
+        response.raise_for_status()
+        self.result = response.content()
```
This error would have to be controlled:
File "/opt/odoo/auto/addons/ai_bridge/models/ai_bridge_execution.py", line 53, in _execute
odoo-1        |     self.result = response.content()
odoo-1        | TypeError: 'bytes' object is not callable

ValentinVinagre commented on ai_bridge/views/ai_bridge_execution.xml:
```diff
@@ -0,0 +1,57 @@
+    <record model="ir.ui.view" id="ai_bridge_execution_tree_view">
+        <field name="model">ai.bridge.execution</field>
+        <field name="arch" type="xml">
+            <tree>
```
Manual creation should not be allowed throughout the model, correct?

ValentinVinagre commented on ai_bridge/static/src/components/chatter_topbar/chatter.esm.js:
```diff
@@ -0,0 +1,22 @@
+                    return;
+                }
+            }
+            await this.env.services.orm.call("ai.bridge", "execute_ai_bridge", [
```
If it throws an error, "_execute" doesn't create the record. Should it create it independently? Would it be better to create it and then execute it? (in the model)

ValentinVinagre reviewed (commented):
Some comments, In this version is complete functional 😄  great work 🚀 .
Tested with personal N8N.

etobella commented:
@HaraldPanten you asked for changes. Can you review?

Abranes commented on ai_oca_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,144 @@
+                continue
+            item = record.env[record.model_id.model].search([], limit=1)
+            if item:
+                record.sample_payload = json.dumps(
+                    record._prepare_payload(item), indent=4
+                )
```
I was testing on Runboat and when I select a date/datetime field I get an error:
![image](https://github.com/user-attachments/assets/0da2b477-156a-4692-9c77-7539bf9cfde4)

Abranes commented on ai_oca_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,95 @@
+        payload = self.ai_bridge_id._prepare_payload(
+            self.env[self.model_id.model].browse(self.res_id), **kwargs
+        )
+        self.payload = payload
```
It might be interesting to save the payload even if it gives an error.

Abranes commented on ai_oca_bridge/views/ai_bridge_execution.xml:
```diff
@@ -0,0 +1,59 @@
+                </sheet>
+            </form>
+        </field>
+    </record>
```
Maybe show the `payload` and the `result`?

ValentinVinagre replied:
These fields are not filled in at the moment, it has been indicated in the review

Abranes reviewed (commented):
Nice one!

Code and functional test on runboat.

Would it be interesting to display the launched executions and their status in the origin record?

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge.xml:
```diff
@@ -0,0 +1,114 @@
+                        <field
+                            name="field_ids"
+                            widget="many2many_tags"
+                            domain="[('model_id', '=', model_id)]"
```
options="{'no_create': True}"

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge.xml:
```diff
@@ -0,0 +1,114 @@
+                        <field name="name" />
+                        <field name="sequence" />
+                        <field name="description" />
+                        <field name="model_id" />
```
options="{'no_create': True}"

ValentinVinagre commented on ai_oca_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,144 @@
+                continue
+            item = record.env[record.model_id.model].search([], limit=1)
+            if item:
+                record.sample_payload = json.dumps(
```
When select fields of type in the field "field_ids":
* date (When it has value)
* datetime (When it has value)
* Binary (When it has value)

give an error when converted to JSON format. From what I've been able to check, the rest of the fields work fine.
![image](https://github.com/user-attachments/assets/ac90224a-a037-425e-86cf-4c5637f1325a)


```
RPC_ERROR
Odoo Server Error
Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 1013, in get
    cache_value = field_cache[record._ids[0]]
KeyError: <NewId origin=1>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/fields.py", line 1161, in __get__
    value = env.cache.get(record, self)
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 1020, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'ai.bridge(<NewId origin=1>,).sample_payload'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1673, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 152, in retrying
    result = func()
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1701, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1905, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 749, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/opt/odoo/auto/addons/web/controllers/dataset.py", line 43, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/auto/addons/web/controllers/dataset.py", line 34, in _call_kw
    return call_kw(Model, method, args, kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 484, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 469, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 6674, in onchange
    todo = [
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 6677, in <listcomp>
    if name not in done and snapshot0.has_changed(name)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 6460, in has_changed
    return self[name] != record[name]
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 6007, in __getitem__
    return self._fields[key].__get__(self, self.env.registry[self._name])
  File "/opt/odoo/custom/src/odoo/odoo/fields.py", line 1222, in __get__
    self.compute_value(recs)
  File "/opt/odoo/custom/src/odoo/odoo/fields.py", line 1404, in compute_value
    records._compute_field_value(self)
  File "/opt/odoo/auto/addons/mail/models/mail_thread.py", line 403, in _compute_field_value
    return super()._compute_field_value(field)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 4276, in _compute_field_value
    fields.determine(field.compute, self)
  File "/opt/odoo/custom/src/odoo/odoo/fields.py", line 98, in determine
    return needle(*args)
  File "/opt/odoo/auto/addons/ai_oca_bridge/models/ai_bridge.py", line 98, in _compute_sample_payload
    record.sample_payload = json.dumps(
  File "/usr/local/lib/python3.10/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/usr/local/lib/python3.10/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/usr/local/lib/python3.10/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/local/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/local/lib/python3.10/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/local/lib/python3.10/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type datetime is not JSON serializable

The above server error caused the following client error:
null
```

ValentinVinagre commented on ai_oca_bridge/static/src/components/chatter_topbar_ai_item/chatter_topbar_ai_item.esm.js:
```diff
@@ -0,0 +1,46 @@
+    }
+    get tooltipInfo() {
+        return {
+            help: markup(this.props.bridge.description || ""),
```
the info div is going to a party when the description is long.

[screen-capture.webm](https://github.com/user-attachments/assets/5e15aed5-0456-4dbf-97b6-e89fa1878b6f)

ValentinVinagre commented on ai_oca_bridge/static/src/components/chatter_topbar/chatter_topbar.xml:
```diff
@@ -0,0 +1,19 @@
+        >
+            <ChatterAITopbar
+                record="chatterTopbar"
+                t-if="chatterTopbar.chatter.webRecord.data.ai_bridge_info !== undefined and chatterTopbar.chatter.webRecord.data.ai_bridge_info.length > 0"
```
Each of the AI ​​Bridge actions can be filtered by domain and/or permission group. It's understood that if the permission group changes, the page must be reloaded.
So, I think it's perfect. When the record is saved, the review is rerun and the corresponding actions are displayed :)

ValentinVinagre commented on ai_oca_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,95 @@
+
+    _name = "ai.bridge.execution"
+    _description = "Ai Execution"  # TODO
+
```
_order = "id desc"

ValentinVinagre replied:
name = fields.Char computed like external identifiers?
![image](https://github.com/user-attachments/assets/23a3324d-7f6b-47fd-bd4f-55bfd7580b9b)

ValentinVinagre commented on ai_oca_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,95 @@
+                **self._execute_kwargs(**kwargs),
+            )
+            response.raise_for_status()
+            self.result = response.content
```
In the test, the execution is generated perfectly. The rest of the fields would need to be filled in:
* payload
* result
* ...

ValentinVinagre commented on ai_oca_bridge/models/ai_bridge_execution.py:
```diff
@@ -0,0 +1,95 @@
+
+    _name = "ai.bridge.execution"
+    _description = "Ai Execution"  # TODO
+
```
Could you log which user performed the action and when, or is the "date" and "user" of creation sufficient?

etobella replied:
Right now I will leave it as it is :wink:

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge_execution.xml:
```diff
@@ -0,0 +1,59 @@
+                    <group>
+                        <field name="ai_bridge_id" />
+                        <field name="res_id" />
+                        <field name="model_id" />
```
add more fields ? 😸

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge_execution.xml:
```diff
@@ -0,0 +1,59 @@
+        <field name="model">ai.bridge.execution</field>
+        <field name="arch" type="xml">
+            <tree create="false" edit="false">
+                <field name="ai_bridge_id" />
```
please add:
* id
* user
* date
* state

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge_execution.xml:
```diff
@@ -0,0 +1,59 @@
+        <field name="model">ai.bridge.execution</field>
+        <field name="arch" type="xml">
+            <search>
+                <field name="ai_bridge_id" />
```
Please add:
search:
* user
* model
* date
* id_record

filter:
* by state.
* date

group by:
* user
* model
* date

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge.xml:
```diff
@@ -0,0 +1,114 @@
+        <field name="arch" type="xml">
+            <tree>
+                <field name="name" />
+                <field name="model_id" />
```
add:
* sequence with widget="handle"
* active

ValentinVinagre reviewed (commented):
some comments.
Would you suggest showing the user a pop-up message in the top right confirming the action has been executed? Currently, there's no user notification, and that could lead to them clicking a lot.
@etobella @HaraldPanten

ValentinVinagre commented on ai_oca_bridge/__manifest__.py:
```diff
@@ -0,0 +1,26 @@
+            "ai_oca_bridge/static/src/**/*.xml",
+            "ai_oca_bridge/static/src/**/*.js",
+        ],
+    },
```
'application': True,

ValentinVinagre commented on ai_oca_bridge/models/ai_bridge.py:
```diff
@@ -0,0 +1,162 @@
+        "res.company",
+        # We leave it empty to allow multiple companies to use the same bridge.
+    )
+    usage = fields.Selection(
```
add one help for this field please

ValentinVinagre commented on ai_oca_bridge/__manifest__.py:
```diff
@@ -0,0 +1,26 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Ai Bridge",
```
Ai Oca Bridge

ValentinVinagre reviewed (commented):
Functionally tested and, in my opinion, ready to be merged. Missing items include:
* points in the header of PR.
* adding a user notification option (same as calendar, etc.)
* fixing the informational popup for user actions when the text is long (roadmap).

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge.xml:
```diff
@@ -0,0 +1,123 @@
+                    />
+                    <group>
+                        <field name="name" />
+                        <field name="sequence" />
```
add company_id please

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge.xml:
```diff
@@ -0,0 +1,123 @@
+                <field name="sequence" widget="handle" />
+                <field name="name" />
+                <field name="model_id" />
+                <field name="usage" />
```
company_id

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge_execution.xml:
```diff
@@ -0,0 +1,84 @@
+                <field name="res_id" />
+                <field name="model_id" />
+                <field name="create_uid" />
+                <field name="create_date" />
```
add company_id

ValentinVinagre commented on ai_oca_bridge/views/ai_bridge_execution.xml:
```diff
@@ -0,0 +1,84 @@
+                        <field name="res_id" />
+                        <field name="model_id" />
+                        <field name="create_uid" />
+                        <field name="create_date" />
```
add company_id

ValentinVinagre reviewed (commented):
Final tests performed (All these tests have been performed with an N8N environment with different flows; Odoo CE multiusers and multicompany):
- When archiving an ai.bridge, the user can continue using it until the user changes or refreshes the screen. Would it be necessary to check if it's archived before launching it? **(I don't see this as necessary for this version.)**
- Same as above, but referring to user groups. If the configuration is changed, even the refresh or screen change doesn't apply. **(I don't see this as necessary for this version.)**
- If the user doesn't have permissions to retrieve information from certain fields, a permissions error will occur. This shouldn't happen with proper permissions settings. However, would it be necessary to execute a sudo to extract data and avoid this potential problem? (A record is not created in the ai.bridge.execution model, so this is correct.) **[Nothing can be done at this point.]**
- **When testing a demo user, you must grant them "Administration/Access Rights." You should extract "id," "_model," and "_id" with sudo, as this is basic information. (This point needs to be addressed.)**
- Test with a demo user by adding the correct "Administration/Access Rights" permissions. Usability is working correctly.**[Nothing can be done at this point.]**
- Correct ai.bridge sorting and filtering in the user widget. Across different users, permissions, and models.**[Nothing can be done at this point.]**
- Connection tests with different correct authentications.**[Nothing can be done at this point.]**
- The connection method change is automatically applied to users, everything is correct.**[Nothing can be done at this point.]**
- If the user is in a multi-company and only has one active, everything works correctly.**[Nothing can be done at this point.]**
- If the user is in multi-company mode and selects View +1 company, all ai.bridges are visible and executable. I believe only those from the active company (the one selected in the company change widget) should be displayed. Another option would be to display the company if it is configured in the ai.bridge configuration (e.g., company1 - magic1). This nomenclature could be a problem if the M2M company field is created in the ai.bridge model. **[It can be added to the Roadmap; it is not considered urgent.]**
- Tests were performed with filtering by company and different executions. The ai.bridge.execution correctly records the execution company.**[Nothing can be done at this point.]**
- The company from which the payload was executed isn't transferred. Would that be necessary? If it is added, it would be "user_company."**[It can be added to the Roadmap]**
- Would it be necessary to be able to select N companies instead of 1 in the ai.bridge model? **[It can be added to the Roadmap]**
- Usage field works correctly.**[Nothing can be done at this point.]**

etobella commented:
> Final tests performed (All these tests have been performed with an N8N environment with different flows; Odoo CE multiusers and multicompany):
> 
> * When archiving an ai.bridge, the user can continue using it until the user changes or refreshes the screen. Would it be necessary to check if it's archived before launching it? **(I don't see this as necessary for this version.)**

Fixed

> * Same as above, but referring to user groups. If the configuration is changed, even the refresh or screen change doesn't apply. **(I don't see this as necessary for this version.)**

Fixed

> * If the user doesn't have permissions to retrieve information from certain fields, a permissions error will occur. This shouldn't happen with proper permissions settings. However, would it be necessary to execute a sudo to extract data and avoid this potential problem? (A record is not created in the ai.bridge.execution model, so this is correct.) **[Nothing can be done at this point.]**
> * **When testing a demo user, you must grant them "Administration/Access Rights." You should extract "id," "_model," and "_id" with sudo, as this is basic information. (This point needs to be addressed.)**

Fixed

> * Test with a demo user by adding the correct "Administration/Access Rights" permissions. Usability is working correctly.**[Nothing can be done at this point.]**
> * Correct ai.bridge sorting and filtering in the user widget. Across different users, permissions, and models.**[Nothing can be done at this point.]**
> * Connection tests with different correct authentications.**[Nothing can be done at this point.]**
> * The connection method change is automatically applied to users, everything is correct.**[Nothing can be done at this point.]**
> * If the user is in a multi-company and only has one active, everything works correctly.**[Nothing can be done at this point.]**
> * If the user is in multi-company mode and selects View +1 company, all ai.bridges are visible and executable. I believe only those from the active company (the one selected in the company change widget) should be displayed. Another option would be to display the company if it is configured in the ai.bridge configuration (e.g., company1 - magic1). This nomenclature could be a problem if the M2M company field is created in the ai.bridge model. **[It can be added to the Roadmap; it is not considered urgent.]**

By default I prefer to leaver records as no multicompany. If the company is selected, everything works as expected.

> * Tests were performed with filtering by company and different executions. The ai.bridge.execution correctly records the execution company.**[Nothing can be done at this point.]**
> * The company from which the payload was executed isn't transferred. Would that be necessary? If it is added, it would be "user_company."**[It can be added to the Roadmap]**

Well, I leave the company of the bridge. Imagine that you have company 1 and 2 selected, but main company is 1 and you execute on a record of company 2.... I prefer to leave the company of the bridge in order to avoid future errors

> * Would it be necessary to be able to select N companies instead of 1 in the ai.bridge model? **[It can be added to the Roadmap]**
> * Usage field works correctly.**[Nothing can be done at this point.]**

ValentinVinagre reviewed (approved):
Well.. Well...It's been a pleasure starting this great project with you, I hope this doesn't stop. Green light, you can now merge, go go 😃 👏🏻 👏🏻

HaraldPanten reviewed (approved):
LGTM

ValentinVinagre commented:
First merge in OCA 😄 thanks @etobella 
/ocabot merge nobump
